zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/IPFSNotebookRepo.java

```diff
@@ -0,0 +1,312 @@
+  private static final String DEFAULT_API_SERVER = "/ip4/127.0.0.1/tcp/5001";
+  private static final Logger LOG = LoggerFactory.getLogger(NotebookRepoSync.class);
+  private ExecutorService executorService = Executors.newCachedThreadPool();
+
+  private Gson gson;
+  private IPFS ipfs;
+  private FileObject ipfsNoteHashesJson;
+  private String encoding;
+  /*
+   *  map of <notedID, List<Base58 hash-string> > i.e Revision History
+   *  { 2A94M5J1Z  : [ QmZQybbanVowHLynssnMjzPJcZ676yB1dA2CpzLN4ZTY48, QmVc.... ],
+   *    2BKZQ2FP6  : [ QmNhPUwuUQ1uD1n22h2CEBFLKPCExCiVc7rcgHmMftmzsv ]
+   *  }
+   */
+  private Map<String, List<Revision>> noteHashes;
+
+  public IPFSNotebookRepo(ZeppelinConfiguration conf) throws IOException {
+    super(conf);
+    encoding = conf.getString(ConfVars.ZEPPELIN_ENCODING);
+    gson = new GsonBuilder().setPrettyPrinting().registerTypeAdapter(Date.class, new
+        NotebookImportDeserializer()).create();
+    String ipfsApiServer = conf.getString("IPFS_API_SERVER",
+        API_SERVER_PROPERTY_NAME, DEFAULT_API_SERVER);
+    ipfs = new IPFS(ipfsApiServer);
+    // creates a ipfsnotehashes.json file in notebook dir if not exists
+    init();
+    // initialize noteHashes Map to load noteID and multihash  from file
+    noteHashes = loadFromFile();
+  }
+
+  /*
+   *  creates a ipfsnotehashes.json file in notebook directory.
+   *  This file will represent the noteHashes Map in Json format.
+   */
+  private void init() throws IOException {
+    FileObject file = getRootDir();
+    ipfsNoteHashesJson = file.resolveFile("ipfsnotehashes.json", NameScope.CHILD);
+    if (!ipfsNoteHashesJson.exists()) {
+      ipfsNoteHashesJson.createFile();
+    }
+  }
+
+  //Reads the ipfsNoteHashesJson file and converts the Json String to Map
+  private Map<String, List<Revision>> loadFromFile() throws IOException {
+    FileContent content = ipfsNoteHashesJson.getContent();
+    InputStream ins = content.getInputStream();
+    String json = IOUtils.toString(ins, encoding);
+    ins.close();
+    if (json.isEmpty() || json == null)
+      return new HashMap<>();
+    Type type = new TypeToken<Map<String, List<Revision>>>() {
+    }.getType();
+    Map<String, List<Revision>> map = gson.fromJson(json, type);
+    return map;
+  }
+
+  /**
+   * Get's the note from peers.
+   *
+   * @param url
+   * @param subject
+   * @return Note
+   * @throws IOException
+   */
+  @Override
+  public Note getNoteFromUrl(String url, AuthenticationInfo subject) throws IOException {
+    //getNote is blocking hence using a timeout
+    final Multihash multihash = new Multihash(Base58.decode(url));
+    Callable<Note> task = new Callable<Note>() {
+      @Override
+      public Note call() throws Exception {
+        return getNote(multihash);
+      }
+    };
+    Future<Note> noteFuture = executorService.submit(task);
+    Note note = null;
+    try {
+      note = noteFuture.get(15, TimeUnit.SECONDS);
```

Comment:
@bzz When the note is not available from peer it should not wait indefinitely so I added timeout in backend. 
And the exception is logged. Should this be propagated to front-end ?
The front-end does not know about the TimeoutException.
@corneadoug should there be another timeout function in front-end ?